clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/provider/view/GoogleLoadBalancerProvider.groovy

```diff
@@ -53,21 +53,32 @@ class GoogleLoadBalancerProvider implements LoadBalancerProvider<GoogleLoadBalan
 
   @Override
   Set<GoogleLoadBalancerView> getApplicationLoadBalancers(String application) {
-    def pattern = Keys.getLoadBalancerKey("*", "*", "${application}*")
-    def identifiers = cacheView.filterIdentifiers(LOAD_BALANCERS.ns, pattern)
-
-    def applicationServerGroups = cacheView.getAll(
+    String pattern = Keys.getLoadBalancerKey("*", "*", "${application}*")
+    Collection<String> identifiers = cacheView.filterIdentifiers(LOAD_BALANCERS.ns, pattern)
+
+    // It is possible to configure a server group in application A to use a load balancer from
+    // application B. Therefore, if (and only if) we have not already retrieved load balancer
+    // identifiers for all applications, we need to retrieve identifiers for every load balancer
+    // associated with a server group from this application.
+    if (application != null && !application.isEmpty()) {
+      Collection<CacheData> applicationServerGroups = cacheView.getAll(
         SERVER_GROUPS.ns,
         cacheView.filterIdentifiers(SERVER_GROUPS.ns, "${GoogleCloudProvider.ID}:*:${application}-*")
-    )
-    applicationServerGroups.each { CacheData serverGroup ->
-      identifiers.addAll(serverGroup.relationships[LOAD_BALANCERS.ns] ?: [])
+      )
+      applicationServerGroups.each { CacheData serverGroup ->
+        Collection<String> relatedLoadBalancers = serverGroup.relationships[LOAD_BALANCERS.ns] ?: []
+        relatedLoadBalancers.each { String lb ->
+          if (!identifiers.contains(lb)) {
```

Comment:
Since `identifiers` is a generic `Collection` we could have up to `O(N)` performance on `contains` which would lead to overall `O(N^2)` performance.  For the redis cache, I think it will always be backed by a `Set` so will have `O(1)` lookup and overall `O(N)` but that's not guaranteed for other providers.

Depending on how long we expect `identifiers` to get, it may well make sense to just create a `Set<String>` out of the collection which would also eliminate the need to check `.contains` at all.

Reply:
Looks like @plumpy beat me to this by 4 minutes...

Reply:
Yeah you two were 100% on the same page! Thank you both for the suggestion, tis implemented!